src/QueryAccess/QueryAccessHandlerBase.php

```diff
@@ -0,0 +1,243 @@
+    return new static(
+      $entity_type,
+      $container->get('entity_type.bundle.info'),
+      $container->get('event_dispatcher'),
+      $container->get('current_user')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getConditions($operation, AccountInterface $account = NULL) {
+    $account = $account ?: $this->currentUser;
+    $entity_type_id = $this->entityType->id();
+    $conditions = $this->buildConditions($operation, $account);
+
+    // Allow other modules to modify the conditions before they are used.
+    $event = new QueryAccessEvent($conditions, $operation, $account);
+    $this->eventDispatcher->dispatch("entity.query_access.{$entity_type_id}", $event);
+
+    return $conditions;
+  }
+
+  /**
+   * Builds the conditions for the given operation and user.
+   *
+   * @param string $operation
+   *   The access operation. One of "view", "update" or "delete".
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   The user for which to restrict access.
+   *
+   * @return \Drupal\entity\QueryAccess\ConditionGroup
+   *   The conditions.
+   */
+  public function buildConditions($operation, AccountInterface $account) {
+    $entity_type_id = $this->entityType->id();
+    if ($account->hasPermission("administer {$entity_type_id}")) {
+      // The user has full access to all operations, no conditions needed.
+      $conditions = new ConditionGroup('OR');
+      $conditions->addCacheContexts(['user.permissions']);
+      return $conditions;
+    }
+
+    if ($this->entityType->entityClassImplements(EntityOwnerInterface::class)) {
+      $entity_conditions = $this->buildEntityOwnerConditions($operation, $account);
+    }
+    else {
+      $entity_conditions = $this->buildEntityConditions($operation, $account);
+    }
+
+    $conditions = NULL;
+    if ($operation == 'view' && $this->entityType->entityClassImplements(EntityPublishedInterface::class)) {
+      $published_key = $this->entityType->getKey('published');
+      if ($entity_conditions) {
+        // Restrict the existing conditions to published entities only.
+        $conditions = new ConditionGroup('AND');
+        $conditions->addCacheContexts(['user.permissions']);
+        $conditions->addCondition($entity_conditions);
+        $conditions->addCondition($published_key, '1');
+      }
+    }
+    else {
+      $conditions = $entity_conditions;
+    }
+
+    if (!$conditions) {
+      // The user doesn't have access to any entities.
+      // Falsify the query to ensure no results are returned.
+      $conditions = new ConditionGroup('OR');
+      $conditions->addCacheContexts(['user.permissions']);
+      $conditions->alwaysFalse();
+    }
+
+    return $conditions;
+  }
+
+  /**
+   * Builds the conditions for entities that have an owner.
+   *
+   * @param string $operation
+   *   The access operation. One of "view", "update" or "delete".
```

Comment:
Can we not limit ourselves to these operations? It would be nice to support an "add-to-cart" operation for example.

Reply:
We should follow core and say "Usually one of". The value is not actually validated anywhere.

Reply:
"Usually one of" would be nice.